### PR TITLE
Breaking up TFRecord parsing logic into smaller blocks @rev mohazahran@

### PR DESCRIPTION
**Pull request by lastmansleeping**
The goal of this PR is to break up the TFRecord parsing function for Example and SequenceExample message types and to add unit tests for the smaller methods. I have tried to consolidate the Example and SequenceExample parsing logic into one consolidated high level flow that is overriden by the sub classes. This enables easy readability and understand the differences between the two.

The main changes are in the following modules:

- [x] `ml4ir/base/data/tfrecord_reader.py`
- [x] `ml4ir/applications/ranking/tests/test_tfrecord_reader.py`
- [x] `ml4ir/applications/classification/tests/test_tfrecord_reader.py`

**Note: I have not modified any of the tests, which is kind of the point of this PR. To break up the code without breaking the tests/functionality.**

In the follow up PR, we will redefine the shape and meaning of sequence features within ml4ir.

Other trivial changes that I probably should have moved to another PR, but I couldn't help myself:
- renaming `ml4ir/applications/ranking/tests/data/config` -> `ml4ir/applications/ranking/tests/data/configs`
- renaming `ml4ir/applications/ranking/tests/data/ranklib_test_data` -> `ml4ir/applications/ranking/tests/data/ranklib`
- corresponding readme/docs updates

### CODE REVIEW

**Review by balikasg — APPROVED** [review]

File: python/ml4ir/applications/classification/tests/test_tfrecord_reader.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+        default_tensor = self.parser.get_default_tensor(
+            self.feature_config.get_feature("user_context"))
+        assert default_tensor.shape == ()
+
+    def test_get_feature(self):
+        """
+        Test fetching feature tensor from extracted feature dictionary
+        """
+        feature_tensor = self.parser.get_feature(self.feature_config.get_feature("query_text"),
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/applications/classification/tests/test_tfrecord_reader.py:82]: It doesn't matter really. I just gave some random dimensions. The goal was to check if the dimensions will be expanded or not.

[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/applications/classification/tests/test_tfrecord_reader.py:82]: I think it will make more sense in the follow up PR. I will set them to more sensible values. In this PR, think of this as also showing that the values don't really matter here.

File: python/ml4ir/base/data/tfrecord_reader.py [hunk 2]
```diff
@@ -10,310 +10,666 @@
...
-    feature_tensor : `tf.Tensor`
-        input feature tensor to be preprocessed
-    feature_info : dict
-        feature configuration for the feature being preprocessed
-    preprocessing_map : `PreprocessingMap` object
-        map of preprocessing feature functions
+    def __init__(self,
+                 feature_config: FeatureConfig,
+                 preprocessing_map: PreprocessingMap,
+                 required_fields_only: bool = False):
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/data/tfrecord_reader.py:24]: Added. You should use `Optional`

File: python/ml4ir/base/data/tfrecord_reader.py [hunk 3]
```diff
@@ -10,310 +10,666 @@
...
+            Returns
+            -------
+            features: dict
+                Dictionary of feature tensors
+            labels: tf.Tensor
+                Label feature tensor used for training
+            """
+            # Parse the proto message to extract feature tensors
+            extracted_features = self.extract_features_from_proto(proto)
+            features_dict = dict()
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/data/tfrecord_reader.py:236]: I am not sure I understand.

File: python/ml4ir/base/data/tfrecord_reader.py [hunk 4]
```diff
@@ -10,310 +10,666 @@
...
+            labels = features_dict.pop(
+                self.feature_config.get_label(key="name"))
+
+            # return X and y which can be used with fit(), predict() and evaluate()
+            return features_dict, labels
+
+        return _parse_fn
+
+
+class ExampleParser(TFRecordParser):
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/data/tfrecord_reader.py:273]: I used this because tensorflow just calls it `Example` and `SequenceExample`. I can change it if you feel strongly though.

File: python/ml4ir/base/data/tfrecord_reader.py [hunk 5]
```diff
@@ -10,310 +10,666 @@
...
+
+    def get_default_tensor(self, feature_info, sequence_size=0):
+        """
+        Get the default tensor for a given feature configuration
+
+        Parameters
+        ----------
+        feature_info: dict
+            Feature configuration information for the feature as specified in the feature_config
+        sequence_size: int, optional
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/data/tfrecord_reader.py:328]: I don't think I can set it to Null as tensorflow can't interpret a null tensor when building the graph. I will try. These might be independent functions to look at, but are actually going to be embedded in the serving signature. So it is limiting what we can do as it has to be tf compatible.

File: python/ml4ir/base/data/tfrecord_reader.py [hunk 6]
```diff
@@ -10,310 +10,666 @@
...
+            if not self.required_fields_only or serving_info.get("required", feature_info["trainable"]):
+                feature_name = feature_info["name"]
+                dtype = feature_info["dtype"]
+                default_value = self.feature_config.get_default_value(
+                    feature_info)
+                if feature_info["tfrecord_type"] == SequenceExampleTypeKey.CONTEXT:
+                    context_features_spec[feature_name] = io.FixedLenFeature(
+                        [], dtype, default_value=default_value
+                    )
+                elif feature_info["tfrecord_type"] == SequenceExampleTypeKey.SEQUENCE:
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/data/tfrecord_reader.py:474]: We can add an else and throw an error if you think that's better.

File: python/ml4ir/base/data/tfrecord_reader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/data/tfrecord_reader.py:273]: Done.

File: python/ml4ir/base/data/tfrecord_reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/data/tfrecord_reader.py:474]: Done.

File: python/ml4ir/base/features/feature_config.py [hunk 7]
```diff
@@ -290,6 +294,23 @@ def get_label(self, key: str = None):
         """
         return self._get_key_or_dict(self.label, key=key)
 
+    def get_mask(self, key: str = None):
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/features/feature_config.py:297]: It is a typical DL mask to identify which records were padded. So for a ranking case, we have query 1 with 5 records and query 2 with 10 records. We pad them to, say 25 records. Now we don't want to use the 20 tail records from query1 and 15 tail records from query2 in the loss and MRR/ACR computation. So we just maintain a binary flag to identify. There are probably better ways to do this using sequence layers. I have not yet explored them.

File: python/ml4ir/applications/classification/tests/test_tfrecord_reader.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    def test_features_spec(self):
+        """
+        Test the feature specification constructed and used to parse the Example proto
+        """
+        features_spec = self.parser.features_spec
+
+        assert isinstance(features_spec, dict)
+
+        # Check if the feature specification matches with the feature_config
+        assert len(set(self.feature_config.get_all_features("name"))
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/applications/classification/tests/test_tfrecord_reader.py:44]: Yeah, this works too. I am just doing in two steps. Can change it.

**Review by mohazahran — APPROVED** [review]
Everything looks good!
